Configurable `!benchmark` comment workflow

samuelburnham opened the pull request:
# `!benchmark` PR comment
Example usage in Lurk:
```
!benchmark --bench fibonacci --bench end2end --features cuda
LURK_RC=100,600,900
LURK_NOISE_THRESHOLD=0.05
```
or simply:
```
!benchmark
```
to run the default benchmarks, features, and runner hardware specified in the caller's workflow file for the repo.

## Implementation details
- GPU usage is first-class: if the `cuda` feature is specified, the workflow will automatically run on a self-hosted runner with the `gpu-bench` label.
- Works on forked PRs, but GPU runners won't work outside for PRs outside the `lurk-lab` organization (see https://github.com/lurk-lab/ci-lab/pull/59#issuecomment-1935293317).

## Succesful runs:
https://github.com/lurk-lab/ci-lab/pull/59#issuecomment-1935289761 and below
https://github.com/lurk-lab/ci-lab/actions/runs/7839337556
https://github.com/lurk-lab/ci-lab/actions/runs/7839353003

huitseeker commented on .github/workflows/bench-pr-comment.yml:
```diff
@@ -1,71 +1,118 @@
-          branchName: ${{ github.ref_name }}
+          printf '${{ github.event.comment.body }}' > comment.txt
+          BENCH_COMMAND=$(head -n 1 comment.txt)
+          echo $BENCH_COMMAND
```
Consider quoting `echo "$BENCH_COMMAND"`, here and elsewhere.

samuelburnham replied:
Could you explain/point me to some resources on why we always want to double-quote env vars with `echo`? Does this extend to all usage of env vars and/or GitHub variables?

Single vs. double quotes and parsing env vars into properly-quoted JSON arrays for the `fromJSON` function was by far the hardest part of this PR, so I'd like to understand it better :pray:

huitseeker replied:
The ultimate reference for me is "Learning the Bash Shell", chapter 8, section command-line processing, and its figure.

huitseeker reviewed (commented):
This looks amazing! Trying to work out details:

> Works on forked PRs, but GPU runners won't work outside for PRs outside the lurk-lab organization

I think you're saying the current PR isn't testable on GPU self-hosted runners until merged, but should be after, correct?

Should we close https://github.com/lurk-lab/lurk-rs/pull/1071 or revamp it to use the shared workflow?

The default for a `!benchmark` command has changed to run on the CPU, and requires saying `--features "cuda"` to use the GPU instead, hasn't it? Can we document that detail in indiscutably spectacular fashion? Make it part of *every* example?

Please assume that folks do *not* read any instructions except maybe those equipped with flashing neon sign and that we'll have to explain to individual after individual why their `!benchmark` command now runs in 10x the time it used to when they said `!gpu-benchmark`.

samuelburnham commented:
> I think you're saying the current PR isn't testable on GPU self-hosted runners until merged, but should be after, correct?

Yes, I also meant that callers outside of `lurk-lab` won't be able to use this with self-hosted runners, not that we recommend such usage anyway due to the Lurk-specific nature of our workflows.
> Should we close https://github.com/lurk-lab/lurk-rs/pull/1071 or revamp it to use the shared workflow?

I think we should close it to preserve the commit history, and open a new one in Lurk and Arecibo once merged
> The default for a !benchmark command has changed to run on the CPU, and requires saying --features "cuda" to use the GPU instead, hasn't it? Can we document that detail in indiscutably spectacular fashion? Make it part of every example?

I've changed added a user note at the top of the workflow and explanation a few places, let me know if you'd like more. I think the change in syntax will be solved through a PSA and more warnings in downstream PRs.

I've also adjusted the default runner to accept a `"self-hosted,gpu-bench"` string in case the caller wants to always use a GPU runner regardless of features. This may be useful for A/B testing the performance of CPU vs. GPU on the same GPU-enabled hardware.

huitseeker reviewed (approved):
Thank you so much! This is amazing progress
